chain/src/main/java/org/ethereum/beacon/chain/GhostHeadFunction.java

```diff
@@ -0,0 +1,222 @@
+
+import org.ethereum.beacon.chain.storage.BeaconBlockStorage;
+import org.ethereum.beacon.chain.storage.BeaconChainStorage;
+import org.ethereum.beacon.chain.storage.BeaconTuple;
+import org.ethereum.beacon.chain.storage.BeaconTupleStorage;
+import org.ethereum.beacon.consensus.HeadFunction;
+import org.ethereum.beacon.consensus.SpecHelpers;
+import org.ethereum.beacon.core.BeaconBlock;
+import org.ethereum.beacon.core.BeaconState;
+import org.ethereum.beacon.core.operations.Attestation;
+import org.ethereum.beacon.core.state.ValidatorRecord;
+import reactor.core.publisher.Flux;
+import tech.pegasys.artemis.ethereum.core.Hash32;
+import tech.pegasys.artemis.util.bytes.Bytes48;
+import tech.pegasys.artemis.util.uint.UInt24;
+import tech.pegasys.artemis.util.uint.UInt64;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * The beacon chain fork choice rule is a hybrid that combines justification and finality with
+ * Latest Message Driven (LMD) Greediest Heaviest Observed SubTree (GHOST). For more info check <a
+ * href="https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#beacon-chain-fork-choice-rule">Beacon
+ * chain fork choice rule</a>
+ */
+public class GhostHeadFunction implements HeadFunction {
+
+  private final BeaconBlockStorage blockStorage;
+  private final BeaconTupleStorage tupleStorage;
+  private final SpecHelpers specHelpers;
+  private final int CHILDREN_SEARCH_LIMIT = Integer.MAX_VALUE;
+  private final Map<Bytes48, Attestation> attestationCache = new HashMap<>();
+  private final Map<UInt64, Set<Bytes48>> validatorSlotCache = new HashMap<>();
+  private BeaconState state;
+
+  public GhostHeadFunction(
+      BeaconChain beaconChain, BeaconChainStorage chainStorage, SpecHelpers specHelpers) {
+    this.tupleStorage = chainStorage.getBeaconTupleStorage();
+    this.blockStorage = chainStorage.getBeaconBlockStorage();
+    this.specHelpers = specHelpers;
+    this.state = tupleStorage.getCanonicalHead().getState();
+    Flux.from(beaconChain.getBlockStatesStream())
+        .doOnNext(beaconTuple -> state = beaconTuple.getState())
+        .subscribe();
+  }
+
+  @Override
+  public Optional<BeaconBlock> update() {
+    Hash32 curHead = blockStorage.getCanonicalHead();
+    UInt64 lastJustifiedSlot = state.getJustifiedSlot();
+    Optional<BeaconTuple> justifiedTuple =
+        blockStorage
+            .getSlotCanonicalBlock(lastJustifiedSlot.getValue())
+            .map(tupleStorage::get)
+            .orElseThrow(() -> new RuntimeException("No justified head found"));
+    BeaconBlock newHead =
+        justifiedTuple
+            .map(this::lmd_ghost)
+            .orElseThrow(() -> new RuntimeException("No justified head found"));
+    if (!newHead.getHash().equals(curHead)) {
+      blockStorage.reorgTo(newHead.getHash());
```

Comment:
IMO, we should get rid of reorgs in block storage and calculate head on the fly after restart. This, also, implies other changes on block storage:
- get rid of canonical chain flat;
- add a state flag with three states: NONE, JUSTIFIED, FINALIZED.